### PR TITLE
feat: Add etcd database support

### PR DESCRIPTION
**Pull request by hudeng-go**
improve concurrent access and high availability of aptly with the help of the characteristics of etcd

Fixes #

## Requirements

All new code should be covered with tests, documentation should be updated. CI should pass.

## Checklist

- [x] unit-test added (if change is algorithm)
- [x] functional test added/updated (if change is functional)
- [x] man page updated (if applicable)
- [x] bash completion updated (if applicable)
- [x] documentation updated
- [x] author name in `AUTHORS`

### CODE REVIEW

**Comment by hudeng-go** [comment]
@lbolla Resolve system test errors at #1022

**Comment by lbolla** [comment]
@hudeng-go Looking good to me. What do you think of this comment on the other ticket? https://github.com/aptly-dev/aptly/pull/1022#issuecomment-1031567248

File: system/run-etcd.py
Comment (lbolla) [inline system/run-etcd.py:1]: This script is almost identical to `system/run.py`:
```
/tmp/aptly etcd-dev ❯ diff system/run.py system/run-etcd.py
97a98,100
>                     t.configOverride = {
>                         "databaseEtcd": "127.0.0.1:2379"
>                     }
```
It would be much better to have a flag or an env variable to tell the test runner script to use etcd, rather than duplicate the whole script.

Moreover, I suspect that using `configOverride` may not work on all cases. Many unittests set their own overrides, so either they will be overwritten with the `databaseEtcd` key or viceversa.

Maybe `databaseEtcd` could come from the environment, e.g. `APTLY_ETCD_DATABASE` and set in `lib.py:BaseTest.config` dictionary?

**Review by lbolla — CHANGES_REQUESTED** [review]

File: system/lib.py
Comment (lbolla) [inline system/lib.py:145]: This will throw a `KeyError` if the env var is not set, which it isn't in the normal `system-tests`. Maybe use `os.environ.get("APTLY_ETCD_DATABASE") or ""` to make sure we have a valid value in the config?

File: context/context.go
Comment (lbolla) [inline context/context.go:291]: I wonder if we should have a more explicit way to pick the database, rather than look at `DatabaseEtcd`.

Maybe having a new config like `databaseBackend` to be one of: "leveldb", "etcd", etc.? What do you think?

**Comment by lbolla** [comment]
@hudeng-go Tests failed... https://github.com/aptly-dev/aptly/runs/5140408342?check_suite_focus=true
They also log *a lot* and take quite a long time, which will eat our free allowance of GH actions... Can you think of a solution?

**Comment by hudeng-go** [comment]
> @hudeng-go Tests failed... https://github.com/aptly-dev/aptly/runs/5140408342?check_suite_focus=true They also log _a lot_ and take quite a long time, which will eat our free allowance of GH actions... Can you think of a solution?

Sorry, it took a long time to link etcd when migrating data. We are looking for the reason.

**Comment by lbolla** [comment]
@hudeng-go I had to stop the CI once again, as it was stuck... Can we make the script so that it would crash at the first error, so we don't continue a clearly failing test?

**Comment by hudeng-go** [comment]
> @hudeng-go I had to stop the CI once again, as it was stuck... Can we make the script so that it would crash at the first error, so we don't continue a clearly failing test?

I have modified the migration script to exit with an error. Sorry again.

**Comment by hudeng-go** [comment]
> 



> @hudeng-go I had to stop the CI once again, as it was stuck... Can we make the script so that it would crash at the first error, so we don't continue a clearly failing test?

In my environment, data can be migrated to normally. I don't know why I can't connect to etcd database in GH action environment.

File: Makefile
Comment (lbolla) [inline Makefile:65]: I guess the second APTLY_DATABASE_TYPE should read APTLY_DATABASE_URL instead?

File: context/context.go
Comment (lbolla) [inline context/context.go:294]: I see the reason for using `DatabaseBackend.Type` as recommended. I wonder though if we can adapt `dbPath` (i.e. `rootPath/db`) into your new `DbConfig` structure? We must be backward compatible though. Maybe it's worth raising a separate ticket once this is merged.

**Comment by lbolla** [comment]
> @hudeng-go I had to stop the CI once again, as it was stuck... Can we make the script so that it would crash at the first error, so we don't continue a clearly failing test?

Once https://github.com/aptly-dev/aptly/pull/1033 is merged, you can rebase to ensure we have stricter timeouts for CI jobs, and avoid tests wasting our GHA allowance.

**Comment by lbolla** [comment]
> > @hudeng-go I had to stop the CI once again, as it was stuck... Can we make the script so that it would crash at the first error, so we don't continue a clearly failing test?
> 
> Once #1033 is merged, you can rebase to ensure we have stricter timeouts for CI jobs, and avoid tests wasting our GHA allowance.

#1033 is merged. Please rebase this branch before carrying out further tests. Thanks!

**Comment by lbolla** [comment]
@hudeng-go On my humble LXD image, which I use for development, `etcd` is using quite a lot of resources. Maybe it crashes on CI because it's too memory/cpu hungry?
Screenshot from 2022-02-14 14-09-04

File: Makefile
Comment (hudeng-go) [reply Makefile:65]: Yes, my mistake.

**Comment by hudeng-go** [comment]
> @hudeng-go On my humble LXD image, which I use for development, `etcd` is using quite a lot of resources. Maybe it crashes on CI because it's too memory/cpu hungry? Screenshot from 2022-02-14 14-09-04

Is aptly in makefile_ DATABASE_ It's my fault that the URL environment variable is incorrectly configured

File: context/context.go
Comment (hudeng-go) [reply context/context.go:294]: I don't quite understand. At present, this dbpath should be available only for leveldb, and the current processing method should be compatible with it. My idea is that if the dbconfig configuration in the configuration file is empty, the default leveldb is used as the back-end, and leveldb compatibility should not be required in dbconfig. This is also explained in the man document.

**Comment by tomsiewert** [comment]
Is there any update to bring this patch into upstream?

**Comment by tomsiewert** [comment]
@randombenj As you seem to be the most active reviewer/contributor for Aptly, could you please review this?

**Comment by neolynx** [comment]
Hi all,

maybe we should rebase this on master first, which seems to be a bit of work.. any chance someone could help with this ?

**Comment by neolynx** [comment]
Hi all,

could anyone have a look at why the test is running into a timeout in https://github.com/aptly-dev/aptly/pull/1284 ?

**Comment by neolynx** [comment]
please git the maintainers force push permissions, so the branch on the PR git repo can be rebased.

In the meanwhile, see: https://github.com/aptly-dev/aptly/pull/1304

**Comment by neolynx** [comment]
Hi all,

it's been a while, but I guess 2 years late is better then never. Please test https://github.com/aptly-dev/aptly/pull/1304 as it is ready to merge soon !

**Pull request closed without merging**
